drivers/console/rtt_console.c

```diff
@@ -18,14 +18,63 @@
 extern void __printk_hook_install(int (*fn)(int));
 extern void __stdout_hook_install(int (*fn)(int));
 
+static bool host_present;
+
+/** @brief Wait for fixed period.
+ *
+ */
+static void wait(void)
+{
+	if (k_is_in_isr()) {
+		k_busy_wait(1000*CONFIG_RTT_TX_RETRY_DELAY_MS);
+	} else {
+		k_sleep(CONFIG_RTT_TX_RETRY_DELAY_MS);
+	}
+}
+
 static int rtt_console_out(int character)
 {
 	unsigned int key;
 	char c = (char)character;
+	unsigned int cnt;
+	int watchdog_counter = CONFIG_RTT_TX_RETRY_CNT;
+
+	do {
+		key = irq_lock();
+		cnt = SEGGER_RTT_WriteNoLock(0, &c, 1);
+		irq_unlock(key);
+
+		/* There are two possible reasons for not writing any data to
+		 * RTT:
+		 * - The host is not connected and not reading the data.
+		 * - The buffer got full and will be read by the host.
+		 * These two situations are distinguished using the following
+		 * algorithm:
+		 * At the beginning, the module assumes that the host is active,
+		 * so when no data is read, it busy waits and retries.
+		 * If, after retrying, the host reads the data, the module
+		 * assumes that the host is active. If it fails, the module
+		 * assumes that the host is inactive and stores that
+		 * information. On next call, only one attempt takes place.
+		 * The host is marked as active if the attempt is successful.
+		 */
+		if (cnt == 0) {
+			if (host_present) {
+				wait();
+				watchdog_counter--;
 
-	key = irq_lock();
-	SEGGER_RTT_WriteNoLock(0, &c, 1);
-	irq_unlock(key);
+				if (watchdog_counter == 0) {
+					host_present = false;
+					break;
+				}
+			} else {
+				break;
+			}
+		} else {
+			/* byte processed - host is present. */
+			host_present = true;
```

Comment:
You never transition to `false` from `true`, so what happens if the Host disconnects? like for example if the user presses Ctrl+C on `JLinkRTTLogger`

Reply:
you will transition to false if cnt equals 0 for number of retries.

Reply:
The logic in this block is really awkward. Is there any way to simplify it? It might help to change the initial condition on line 63 to `if (cnt > 0)` and handle that case first, then move on to the more complicated logic.